### PR TITLE
Fix .paint_path bug noted in issue #473

### PR DESCRIPTION
**Pull request by jsvine**
See #473 for details. Focuses on the handling of non-rect quadrilaterals, the decomposition of complex (m.*h)* paths into subpaths, and assigning those subpaths the correct LTCurve/LTRect type.

**Pull request**

Thanks for improving pdfminer.six! Please include the following information to
help us discuss and merge this PR:

__A description of why this PR is needed. What does it fix? What does it improve?__

Please see #473 for details.

__A summary of the things that this PR changes.__

This PR partially rewrites `PDFLayoutAnalyzer.paint_path(...)`, and also adds a test to `tests/test_converter.py`.

**How Has This Been Tested?**

See changes to `tests/test_converter.py`, with more context available in #473 

**Checklist**

- [X] I have added tests that prove my fix is effective or that my feature 
  works
- [ ] (not applicable) I have added docstrings to newly created methods and classes
- [X] I have optimized the code at least one time after creating the initial 
  version
- [ ] (not applicable) I have updated the [README.md](../README.md) or I am verified that this
  is not necessary
- [ ] (not applicable) I have updated the [readthedocs](../docs/source) documentation or I 
  verified that this is not necessary
- [X] I have added a consice human-readable description of the change to 
  [CHANGELOG.md](../CHANGELOG.md)

### CODE REVIEW

File: pdfminer/converter.py [hunk 1]
```diff
@@ -87,9 +95,10 @@ def paint_path(self, gstate, stroke, fill, evenodd, path):
                 line = LTLine(gstate.linewidth, (x0, y0), (x1, y1), stroke,
                               fill, evenodd, gstate.scolor, gstate.ncolor)
                 self.cur_item.add(line)
+                return
```
Comment (pietermarsman) [inline pdfminer/converter.py:98]: Use if-elif-else statements instead of early returns

File: pdfminer/converter.py [hunk 2]
```diff
@@ -77,8 +75,18 @@ def render_image(self, name, stream):
...
         """Paint paths described in section 4.4 of the PDF reference manual"""
         shape = ''.join(x[0] for x in path)
+
+        # if path contains multiple subpaths, split them up and reprocess
+        if shape.count("m") > 1:
+            m_indices = [i for i, x in enumerate(shape) if x == "m"]
+            m_indices_zipped = zip(m_indices, m_indices[1:] + [None])
+            subpaths = [path[a:b] for a, b in m_indices_zipped]
+            for sp in subpaths:
+                self.paint_path(gstate, stroke, fill, evenodd, sp)
```
Comment (pietermarsman) [inline pdfminer/converter.py:85]: Can you try to make this code a bit shorter? E.g. with something along the lines of:

```python
>>> for m in re.finditer(r'ml+h', 'mlllhmlllh'):
...     print(m.start(0), m.end(0))
(0, 5)
(5, 10)
```

File: tests/test_converter.py [hunk 3]
```diff
@@ -32,6 +32,69 @@ def test_paint_path_multiple_mlllh(self):
...
+            ("h",),
+        ]), [LTRect])
+
+        assert_equal(get_types([
+            ("m", 110, 90),
+            ("l", 190, 10),
+            ("l", 190, 90),
+            ("l", 110, 10),
+            ("h",),
+        ]), [LTCurve])
```
Comment (pietermarsman) [inline tests/test_converter.py:61]: Nice test! :)

**Review by pietermarsman — CHANGES_REQUESTED** [review]

**Comment by pietermarsman** [comment]
@jsvine I think the code looks good. The tests look really good :). I suggested some minor changes.

File: pdfminer/converter.py (hunk 2 above)
Comment (jsvine) [reply pdfminer/converter.py]: Clever! Adjusted in latest push. Opted for `r'm[^m]+'` as the pattern, to catch subpaths that might not exclusively use `l` (e.g., `c` for cubic Bezier paths). Not sure exactly how those are parsed in pdfminer, though; let me know if you foresee problems with the pattern.

File: pdfminer/converter.py (hunk 1 above)
Comment (jsvine) [reply pdfminer/converter.py]: Sure, done! Required some reordering of the logic, but in a good way (I think?).

File: pdfminer/converter.py (hunk 2 above)
Comment (pietermarsman) [reply pdfminer/converter.py]: This regex pattern looks good! :)

File: pdfminer/converter.py (hunk 1 above)
Comment (pietermarsman) [reply pdfminer/converter.py]: I reorderdered the if-selse staements once more. Let me know what you think.

File: pdfminer/converter.py [hunk 4]
```diff
@@ -76,49 +76,46 @@ def paint_path(self, gstate, stroke, fill, evenodd, path):
...
+                    curve = self._create_curve(gstate, stroke, fill, evenodd, path)
+                    self.cur_item.add(curve)
+
+            else:
+                curve = self._create_curve(gstate, stroke, fill, evenodd, path)
+                self.cur_item.add(curve)
+
+    def _create_curve(self, gstate, stroke, fill, evenodd, path):
+        """Create a `LTCurve` object for the paint path operator"""
+        pts = [apply_matrix_pt(self.ctm, point) for p in path for point in zip(p[::2], p[1::2])]
```
Comment (jsvine) [inline pdfminer/converter.py:116]: `tests/test_converter.py:test_paint_path_quadrilaterals` appears to be failing on this line.

**Comment by jsvine** [comment]
Thanks, @pietermarsman. The reordering of the logic makes general sense to me, but `tests/test_converter.py:test_paint_path_quadrilaterals` now fails (see above).

File: pdfminer/converter.py (hunk 4 above)
Comment (pietermarsman) [reply pdfminer/converter.py:116]: I'll look into it

Comment (pietermarsman) [reply pdfminer/converter.py]: I made an error when restructuring the for loops into a list comprehension. It should be `zip(p[1::2], p[2::2])`.

Comment (pietermarsman) [reply pdfminer/converter.py]: The new tests are already proving their value :D

**Review by pietermarsman — APPROVED** [review]

**Pull request merged**